DragaliaAPI/Controllers/Dragalia/FortController.cs

```diff
@@ -43,13 +70,695 @@ await this.fortRepository.GetBuilds(this.DeviceAccountId).ToListAsync()
+                result = 1,
+                build_id = request.build_id,
+                fort_bonus_list = bonusList,
+                production_rp = StubData.ProductionRp,
+                production_st = StubData.ProductionSt,
+                production_df = StubData.ProductionDf,
+                fort_detail = fortDetail,
+                update_data_list = updateDataList,
+            };
+        return this.Ok(data);
+    }
+
+    [HttpPost("build_start")]
+    public async Task<DragaliaResult> BuildStart(FortBuildStartRequest request)
+    {
+        DbPlayerUserData userData = await this.userDataRepository
+            .GetUserData(this.DeviceAccountId)
+            .FirstAsync();
+
+        IQueryable<DbPlayerMaterial> userMaterials = this.inventoryRepository
+            .GetMaterials(this.DeviceAccountId);
+
+        FortDetail fortDetail = this.fortRepository.Details.Select(mapper.Map<FortDetail>).First();
+
+        // Check Carpenter available
+        if (fortDetail.working_carpenter_num > fortDetail.carpenter_num)
+        {
+            throw new DragaliaException(
+                ResultCode.FortBuildCarpenterBusy,
+                $"All carpenters are currently busy"
+            );
+        }
+
+        // Get build plans
+        FortPlants BuildPlantId = (FortPlants)request.fort_plant_id;
+        int buildIdAtLevel1 = MasterAssetUtils.GetPlantDetailId(BuildPlantId, 1);
+        FortPlantDetail plantDetail = MasterAsset.FortPlant.Get(buildIdAtLevel1);
+
+        // Remove player resources
+        await ConsumePlayerMaterials(userMaterials, userData, plantDetail);
+
+        // Start building
+        DateTime startDate = DateTime.UtcNow;
+        DateTime endDate = startDate.AddSeconds(plantDetail.Time);
+
+        DbFortBuild build = new()
+        {
+            DeviceAccountId = this.DeviceAccountId,
+            PlantId = BuildPlantId,
+            Level = 1,
+            PositionX = request.position_x,
+            PositionZ = request.position_z,
+            BuildStartDate = startDate,
+            BuildEndDate = endDate,
+            IsNew = true,
+            LastIncomeDate = DateTimeOffset.UnixEpoch
+        };
+        await this.fortRepository.AddBuild(build);
+
+        // Increment worker carpenters
+        await IncrementCarpenterUsage(fortDetail);
+
+        UpdateDataList updateDataList = this.updateDataService.GetUpdateDataList(
+            this.DeviceAccountId
+        );
+
+        await this.fortRepository.SaveChangesAsync();
+
+        int startDateUnix = (int)((DateTimeOffset)startDate).ToUnixTimeSeconds();
+        int endDateUnix = (int)((DateTimeOffset)endDate).ToUnixTimeSeconds();
+        FortBuildStartData data =
+            new()
+            {
+                result = 1,
+                build_id = (ulong)build.BuildId,
+                build_start_date = startDateUnix,
+                build_end_date = endDateUnix,
+                remain_time = endDateUnix - startDateUnix,
```

Comment:
`remain_time` expects an `int`, so the alternative using your suggestion would be something like:

`(int)(endDate - startDate).TotalSeconds`

Also worth noting that we would still need to have the `startDateUnix` and `endDateUnix`, as `build_start_date` and `build_end_date` also require `int`s.

Don't think this change would make the code better or easier to read in this instance, unless I am missing something!

Reply:
there is a custom TimestampToUnixInt formatter for the request/response models (see https://github.com/SapiensAnatis/DragaliaAPI/blob/develop/DragaliaAPI/MessagePack/TimespanToUnixIntFormatter.cs), so by changing the type to a TimeSpan it gets automatically converted to and from an int during serialization.

Reply:
Note that the same also applies to both startDateUnix and endDateUnix, as those can be changed to DateTimeOffsets which are converted using the custom DateTimeOffsetToUnixInt converter (see https://github.com/SapiensAnatis/DragaliaAPI/blob/develop/DragaliaAPI/MessagePack/DateTimeOffsetToUnixIntFormatter.cs)

Reply:
Oh, sorry I wasn't clear, I also meant you should change the type in the response to `TimeSpan` and it will get serialized as an int (so the game will never know). The types autogenerated and we frequently change them to make them easier to use because a lot of the game's datatypes make no sense (e.g. 0/1 for bools)

Reply:
Ah, that makes a lot of sense now, thanks! I'll make those changes